InterruptClient.rb

```diff
@@ -0,0 +1,230 @@
+    'magenta',
+    'cyan',
+    'blue',
+    'light_green',
+  ]
+
+  def initialize(server_host, server_port)
+    @server_host = server_host
+    @server_port = server_port
+
+    @client = UDPSocket.new
+    @client.connect(@server_host, @server_port)
+
+    @latest_chat = '' # string timestamp of latest chat received
+  end
+
+  def run
+    set_name
+    handshake
+    instructions
+    terminal_config
+    receive_loop
+  end
+
+  def set_name
+    puts START_MSG
+    print PROMPT
+    @name = STDIN.gets.chomp
+  end
+
+  def handshake
+    ackd = false
+    while not ackd
+      send_msg(msg_connect(@name))
+      incoming = IO.select([@client], nil, nil, HANDSHAKE_WAIT)
+      if ((not incoming.nil?) && ack?)
+        ackd = true
+      end
+    end
+  end
+
+  def ack?
+    data, sender = @client.recvfrom(MAX_MSG_LENGTH)
+    msg = parse_msg(data, sender)
+    return false if msg.nil?
+
+    msg['type'] == 'ack'
+  end
+
+  def instructions
+    print Console.clear
+    puts INSTRUCTIONS + "\n" * 12
+  end
+
+  def receive_loop
+    loop do
+      to_read = IO.select([@client, STDIN]) # check what's ready to read: socket, or terminal input
+      to_read[0].each { |ios|
+        if ios == @client
+          msg, sender = ios.recvfrom(MAX_MSG_LENGTH)
+          handle_msg(msg, sender)
+        elsif ios.tty? # ios comes from terminal
+          input = STDIN.getc
+          handle_key(input)
+        end
+      }
+    end
+  end
+
+  def handle_msg(msg, sender)
+    msg = parse_msg(msg, sender)
+    return if msg.nil?
+
+    case msg['type']
+    when 'chat'
+      time = msg['time']
+      return if time < @latest_chat
+
+      @latest_chat = time
+      chat_array = msg['body']
+      names_array = msg['names']
+
+      names_string, names_length = names_data(names_array)
+
+      string = Console.left(chat_array.length) + Console.up(10) +
+          names_string +
+          Console.left(names_length) + Console.down(10) +
+          chat_string(chat_array)
+
+      print string
+    end
+  end
+
+  def chat_string(chat_array)
+    chat_array.inject('') { |string, item|
+        text, color_code = item
+        color = color_code.nil? ? nil : COLORS[color_code]
+
+        string + Console.color(color, text)
+      }
+  end
+
+  def names_data(names_array)
+    names_string = ''
+    names_length = 0 # names_string may have color encoding chars, so we can't just check its length
+
+    names_array.each {|item|
+      nickname, color_code, emph = item
+      names_length += nickname.length + 2
+
+      nickname = color_code.nil? ? nickname : Console.color(COLORS[color_code], nickname)
+      nickname = emph ? Console.emph(nickname) : nickname
+      names_string += " #{nickname} "
+      }
+
+    return [names_string, names_length]
+  end
+
+  # disregard eg return, delete, backspace keys presses
+  # only pass on to server word, punctuation, or space characters
+  # TODO: strip out two chars following '\e'
+  def handle_key(input)
```

Comment:
I recommend just trapping SIGINT . Then we can just exit by hitting ctrl-c https://gist.github.com/sauloperez/6592971